module.go

```diff
@@ -0,0 +1,268 @@
+package tfaddr
+
+import (
+	"fmt"
+	"path"
+	"regexp"
+	"strings"
+
+	svchost "github.com/hashicorp/terraform-svchost"
+)
+
+var moduleSourceLocalPrefixes = []string{
+	"./",
+	"../",
+	".\\",
+	"..\\",
+}
+
+func isModuleSourceLocal(raw string) bool {
+	for _, prefix := range moduleSourceLocalPrefixes {
+		if strings.HasPrefix(raw, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+// ModuleSourceRegistry is representing a module listed in a Terraform module
+// registry.
+type ModuleSourceRegistry struct {
+	// PackageAddr is the registry package that the target module belongs to.
+	// The module installer must translate this into a ModuleSourceRemote
+	// using the registry API and then take that underlying address's
+	// PackageAddr in order to find the actual package location.
+	PackageAddr ModuleRegistryPackage
+
+	// If Subdir is non-empty then it represents a sub-directory within the
+	// remote package that the registry address eventually resolves to.
+	// This will ultimately become the suffix of the Subdir of the
+	// ModuleSourceRemote that the registry address translates to.
+	//
+	// Subdir uses a normalized forward-slash-based path syntax within the
+	// virtual filesystem represented by the final package. It will never
+	// include `../` or `./` sequences.
+	Subdir string
+}
+
+// DefaultModuleRegistryHost is the hostname used for registry-based module
+// source addresses that do not have an explicit hostname.
+const DefaultModuleRegistryHost = svchost.Hostname("registry.terraform.io")
+
+var moduleRegistryNamePattern = regexp.MustCompile("^[0-9A-Za-z](?:[0-9A-Za-z-_]{0,62}[0-9A-Za-z])?$")
+var moduleRegistryTargetSystemPattern = regexp.MustCompile("^[0-9a-z]{1,64}$")
+
+// ParseRawModuleSourceRegistry only accepts module registry addresses, and
+// will reject any other address type.
+func ParseRawModuleSourceRegistry(raw string) (ModuleSourceRegistry, error) {
+	// Before we delegate to the "real" function we'll just make sure this
+	// doesn't look like a local source address, so we can return a better
+	// error message for that situation.
+	if isModuleSourceLocal(raw) {
+		return ModuleSourceRegistry{}, fmt.Errorf("can't use local directory %q as a module registry address", raw)
+	}
```

Comment:
At the risk of drifting a little bit from the (non-public) TF Core implementation, is it worth also removing this check here?

I expect downstream consumers would just check for local addresses 1st anyway.

We can be more defensive if we find edge cases where the parser misinterprets addresses. In this case though it looks like we'd just get slightly different error, which seems okay?

```
--- FAIL: TestParseRawModuleSourceRegistry (0.00s)
    --- FAIL: TestParseRawModuleSourceRegistry/local_path_from_current_dir (0.00s)
        module_test.go:219: wrong error messages
            got:  a module registry source address must have either three or four slash-separated components
            want: can't use local directory "./boop" as a module registry address
    --- FAIL: TestParseRawModuleSourceRegistry/local_path_from_parent_dir (0.00s)
        module_test.go:219: wrong error messages
            got:  a module registry source address must have either three or four slash-separated components
            want: can't use local directory "../boop" as a module registry address
FAIL
FAIL	github.com/hashicorp/terraform-registry-address	0.899s
FAIL
```

Reply:
Good idea, I removed the check and the `isModuleSourceLocal` function.